src/apps/virtio_net/virtio_net_dev.lua

```diff
@@ -0,0 +1,208 @@
+module(..., package.seeall)
+
+local debug = _G.developer_debug
+
+local ffi       = require("ffi")
+local C         = ffi.C
+local S         = require('syscall')
+local pci       = require("lib.hardware.pci")
+local bit       = require('bit')
+local gvring    = require('apps.virtio_net.guest_vring')
+require("lib.checksum_h")
+require('lib.virtio.virtio_h')
+
+local band, bor, rshift = bit.band, bit.bor, bit.rshift
+
+local function fstruct(def)
+   local struct = {}
+   local offset = 0
+   for ct, fld in def:gmatch('([%a_][%w_]*)%s+([%a_][%w_]*);') do
+      ct = ffi.typeof(ct)
+      struct[fld] = {
+         fieldname = fld,
+         ct = ct,
+         size = ffi.sizeof(ct),
+         offset = offset,
+      }
+      offset = offset + struct[fld].size
+   end
+   return struct, offset
+end
+
+local function fieldrd(field, fd)
```

Comment:
This is an adapter to be able to read/write to a non-mmap()'able PCI resource, especially since it requires that the operations are in the exact size of the field.  I don't know if there are other similar cases.